subsys/bluetooth/services/cgms/cgms_racp.c

```diff
@@ -0,0 +1,473 @@
+			rc = report_num_recs_handler(task->peer,
+				&task->req_data[1], task->req_len - 1);
+			break;
+		default:
+			rc = generic_handler(task->peer, task->req_data[0],
+					RACP_RESPONSE_OPCODE_UNSUPPORTED);
+			break;
+		}
+		k_mutex_unlock(&lock);
+	} else {
+		rc = -EBUSY;
+	}
+}
+
+int cgms_racp_recv_request(struct bt_conn *peer, const uint8_t *req_data, uint16_t req_len)
+{
+	int rc;
+
+	/* Early check of the data length. */
+	if (req_data == NULL || req_len == 0) {
+		LOG_INF("RACP: Empty command");
+		return -ENODATA;
+	}
+
+	/* All tasks are executed in sys q expect abortion.
+	 * So we need to handler abortion request here.
+	 */
+	if (req_data[0] == RACP_OPCODE_ABORT_OPERATION) {
+		if (req_len > 1 && req_data[1] == RACP_OPERATOR_NULL) {
+			rc = k_work_cancel(&report_record_task.item);
+			if (rc == 0) {
+				rc = generic_handler(peer, RACP_OPCODE_ABORT_OPERATION,
+				RACP_RESPONSE_SUCCESS);
+			} else {
+				rc = generic_handler(peer, RACP_OPCODE_ABORT_OPERATION,
+				RACP_RESPONSE_ABORT_FAILED);
+			}
+		} else {
+			rc = generic_handler(peer, RACP_OPCODE_ABORT_OPERATION,
+				RACP_RESPONSE_INVALID_OPERATOR);
+		}
+		LOG_INF("RACP: abortion %s", rc == 0 ? "done" : "failed");
+		return rc;
+	}
+
+	/* For other requests, prepare the data and submit to workqueue. */
+	report_record_task.peer = peer;
+	memcpy(report_record_task.req_data, req_data, req_len);
+	report_record_task.req_len = req_len;
+	rc = k_work_submit_to_queue(&racp_work_q, &report_record_task.item);
+	LOG_INF("RACP: work submission %s", rc > 0 ? "done" : "failed");
+	return rc;
+}
+
+int cgms_racp_meas_add(struct cgms_meas meas)
+{
+	int rc;
+
+	if (k_mutex_lock(&lock, K_NO_WAIT) == 0) {
+		/* If there is free slot, use it and increase count.
+		 * Otherwise, the count remains max, and the oldest
+		 * record pointed by head is overwritten, and increase
+		 * head by 1.
+		 */
+		if (database.count < ARRAY_SIZE(database.meas)) {
+			memcpy(&database.meas[(database.head + database.count)
+				% ARRAY_SIZE(database.meas)],
+				&meas, sizeof(meas));
+			database.count++;
+		} else {
+			memcpy(&database.meas[(database.head + database.count)
+				% ARRAY_SIZE(database.meas)],
+				&meas, sizeof(meas));
+			database.head = (database.head + 1) % ARRAY_SIZE(database.meas);
+		}
+		LOG_DBG("Database head [%d], [%d]/[%d] used.",
+			database.head, database.count, ARRAY_SIZE(database.meas));
+		k_mutex_unlock(&lock);
+		rc = 0;
+	} else {
+		rc = -EBUSY;
+	}
+
+	return rc;
+}
+
+int cgms_racp_meas_get_latest(struct cgms_meas *meas)
+{
+	int rc;
+
+	if (k_mutex_lock(&lock, K_NO_WAIT) == 0) {
+		if (database.count > 0) {
+			memcpy(meas,
+				&database.meas[(database.head + database.count - 1)
+						% ARRAY_SIZE(database.meas)],
+				sizeof(struct cgms_meas));
+			rc = 0;
+		} else {
+			rc = -ENODATA;
+		}
+		k_mutex_unlock(&lock);
+	} else {
+		rc = -EBUSY;
+	}
+
+	return rc;
+}
+
+void cgms_racp_init(void)
+{
+	memset(&database, 0, sizeof(database));
+
+	k_mutex_init(&lock);
```

Comment:
Is it possible to do all operations on a record database in a workqueue thread? Then a mutex would be unnecessary.

Reply:
I think it is not esay to eliminate. IMO, a mutex, or a similar thing, is unavoidable. For example, to add record into db, if we do not share database between two threads, we will share the record. Thread safety is always a concern as long as we pass pointers (i.e., shared memory). A kernel service that is pass-by-value would help (say, socket), but it seems too heavy for RACP.